cocotb/runner.py

```diff
@@ -0,0 +1,694 @@
+                )
+            )
+            if self.vhdl_sources:
+                self.env["GPI_EXTRA"] = cocotb.config.lib_name_path("vhpi", "riviera") + ":cocotbvhpi_entry_point"
+
+        if self.waves:
+            do_script += "log -recursive /*;"
+
+        do_script += "run -all \nexit"
+
+        do_file = tempfile.NamedTemporaryFile(delete=False)
+        do_file.write(do_script.encode())
+        do_file.close()
+
+        return [["vsimsa"] + ["-do"] + ["do"] + [do_file.name]]
+
+
+class Verilator(Simulator):
+    def __init__(self, *argv, **kwargs):
+        super(Verilator, self).__init__(*argv, **kwargs)
+
+        self.env["CXXFLAGS"] = self.env.get("CXXFLAGS", "") + " -std=c++11"
+
+    def get_include_commands(self, includes):
+        include_cmd = []
+        for dir in includes:
+            include_cmd.append("-I" + dir)
+
+        return include_cmd
+
+    def get_define_commands(self, defines):
+        defines_cmd = []
+        for define in defines:
+            defines_cmd.append("-D" + define)
+
+        return defines_cmd
+
+    def get_parameter_commands(self, parameters):
+        parameters_cmd = []
+        for name, value in parameters.items():
+            parameters_cmd.append("-G" + name + "=" + str(value))
+
+        return parameters_cmd
+
+    def build_command(self):
+
+        if self.vhdl_sources:
+            raise ValueError("This simulator does not support VHDL")
+
+        if self.hdl_topmodule is None:
+            raise ValueError("This simulator requires hdl_topmodule parameter to be specified")
+
+        cmd = []
+
+        verilator_cpp = os.path.join(os.path.dirname(os.path.dirname(cocotb.config.libs_dir)), "share", "verilator.cpp")
+        verilator_cpp = os.path.join(os.path.dirname(cocotb.__file__), "share", "lib", "verilator", "verilator.cpp")
+
+        verilator_exec = shutil.which("verilator")
+        if verilator_exec is None:
+            raise ValueError("Verilator executable not found.")
+
+        cmd.append(
+            [
+                "perl",
+                verilator_exec,
+                "-cc",
+                "--exe",
+                "-Mdir",
+                self.build_dir,
+                "-DCOCOTB_SIM=1",
+                "--top-module",
+                self.hdl_topmodule,
+                "--vpi",
+                "--public-flat-rw",
+                "--prefix",
+                "Vtop",
+                "-o",
+                self.hdl_topmodule,
+                "-LDFLAGS",
+                "-Wl,-rpath,{LIB_DIR} -L{LIB_DIR} -lcocotbvpi_verilator".format(LIB_DIR=cocotb.config.libs_dir),
+            ]
+            + self.compile_args
+            + self.get_define_commands(self.defines)
+            + self.get_include_commands(self.includes)
+            + self.get_parameter_commands(self.parameters)
+            + [verilator_cpp]
+            + self.verilog_sources
+        )
+
+        cmd.append(["make", "-C", self.build_dir, "-f", "Vtop.mk"])
```

Comment:
in verilator you can set a few options during this make command like: 
```
make OPT_SLOW="-O3" OPT_FAST="-O3" VM_PARALLEL_BUILDS=1 -j8 -C sim_build/ -f Vtop.mk
```

Is there a possibility of exposing these to the user ?

Reply:
For verilator, one can add an extra argument to test like `make_args`. It will break compatibility but ... 🤷🏻

Reply:
I don't see how it will break compatibility? Other simulators would just ignore or throw a warning if the argument is given.

Reply:
>I don't see how it will break compatibility? Other simulators would just ignore or throw a warning if the argument is given.

We can `make_args` to all simulators  `Simulator` class or only for `Verilator` class. It will break if we do `Verilator` class.